plugins/steve/resource-class.ts

```diff
@@ -718,15 +761,15 @@ export default class Resource {
 
   // ------------------------------------------------------------------
 
-  hasAction(actionName) {
+  hasAction(actionName: string) {
     return !!this.actionLinkFor(actionName);
   }
 
-  actionLinkFor(actionName) {
+  actionLinkFor(actionName: string) {
     return (this.actions || this.actionLinks || {})[actionName];
   }
 
-  doAction(actionName, body, opt = {}) {
+  doAction(actionName: string, body: any, opt = {} as any) {
```

Comment:
We could do something funcky here with 'generics'

```
  doAction<T>(actionName: string, body: any, opt: T) {
    return this.$dispatch('resourceAction', {
      resource: this,
      actionName,
      body,
      opt: opt || {},
    });
  }
```
Which would mean anyone using this action could specify a type relevant to the place where it's used

Reply:
You do not need to cast the `T` type as it gets automatically inferred by `opt`.
It would still required the returned value from the method and the type for `body` though, which may be `K` (or any letter) if is something we dont want to type. If instead is a resource, we should be able to reference itself in somehow within the class or the extended resource class.

Reply:
Ah, good point, typing body this way would be much more helpful

Reply:
Oh sorry, I remembered it wrong, you have to put the `T` but then it get inferred and does not need to be cast on call :grimacing: 

I have something similar for the utilities and I've to state as follow:

```ts
export function removeObject<T>(ary: T[], obj: T): T[] {
  const idx = ary.indexOf(obj);

  if ( idx >= 0 ) {
    ary.splice(idx, 1);
  }

  return ary;
}

removeObject( [ ], { foo: 'bar' } ) // --> { foo: string; }[]
```

Reply:
I'm finding that the problem with T is that it won't let me set a default value for the param of type T.

```js
doAction<T>(actionName: string, body: T, opt: T = {}): void {
    return this.$dispatch('resourceAction', {
      resource: this,
      actionName,
      body,
      opt,
    });
  }
```

> Type '{}' is not assignable to type 'T'.  'T' could be instantiated with an arbitrary type which could be unrelated to '{}'.